### PR TITLE
Make supported course formats configurable

### LINKED ISSUE #34: Process for course format support
@PhMemmel  Just opened this as a place to have this discussion.  I had an idea about a block setting that would list all course formats installed and allow the site owner to select the course formats that the mass actions block would be available.  We can preselect formats that are "known" working formats like topcoll is currently.  But this would also allow sites to select formats that we are either unaware of or are only partially supported without us having to manage it in code.  Just a thought.

### PR DESCRIPTION
**Pull request by PhMemmel**
Closes #34 

@Syxton How should we deal with this new option in behat tests? Anything else required there? Feel free to just push on top of this branch :-)

### CODE REVIEW

**Comment by Syxton** [comment]
@PhMemmel  Adding an upgrade.php file.  We have never needed one before, but we will with the new admin setting.

If I install a new version of massactions that now supports "tiles" for instance.  I would want my settings updated to check the supported formats, keeping my other alterations.  In the same way, if we stop supporting a plugin, by default I would want to uncheck that setting.  I have added the methods to make that very simple on a version update.  The only case it doesn't fix is the case of a user installing a supported format like "topcoll" after the massactions block setting was already made and modified.

**Comment by PhMemmel** [comment]
@Syxton Good catch, but where are the functions in `upgrade.php` called from? Cannot find where this happens?

**Comment by Syxton** [comment]
@PhMemmel They aren't currently used anywhere.  But as soon as we add a course format to our "supported" list or remove one from the supported list, they will be.

**Comment by Syxton** [comment]
@PhMemmel I'm thinking maybe we should lock in supported formats.  Only non-supported formats can be toggled on/off.  So our officially supported formats don't show up in the list (or they do show up in the list but are not editable)

**Comment by PhMemmel** [comment]
I had the same thought, but decided against at first. I liked the idea of the admin being able to disable the block in certain course formats, for example if a custom theme has some complications with a certain course format? An Admin would have the opportunity to disable it for the time being in this case. What do you think?

**Comment by Syxton** [comment]
@PhMemmel Yeah, I also don't like the idea of giving only partial control to a site owner.  It's not ideal, but the worst case scenario is that they have to change the setting to re-select a fully supported format.  Ok, we will keep it as is.  I think I like this change.  I haven't yet looked at the testing.  The core tests all work because of the default settings.  New supported formats will alter the default so at least in my head they should also work without alteration.

File: lang/en/block_massaction.php [hunk 1]
```diff
@@ -65,6 +65,9 @@
 $string['noactionsavailable'] = 'You do not have the permissions to execute any of the possible operations this block is providing';
 $string['nomovingtargetselected'] = 'Please select a target section';
 $string['sectionnotexist'] = 'Target section does not exist';
+$string['applicablecourseformats'] = 'Applicable course formats';
+$string['applicablecourseformats_description'] = 'Select for which course formats the Mass Actions block should be available for the users. '
```
Comment (Syxton) [inline lang/en/block_massaction.php:69]: Select for which course formats the Mass Actions block should be available for the users.
to
Mass Actions block will only be available for the selected course formats.

**Review by Syxton — COMMENTED** [review]
I think I'm on board for this change.  The only issue was a little bit of a clunky description for the new setting.

**Comment by PhMemmel** [comment]
> @PhMemmel Yeah, I also don't like the idea of giving only partial control to a site owner. It's not ideal, but the worst case scenario is that they have to change the setting to re-select a fully supported format. Ok, we will keep it as is. I think I like this change. I haven't yet looked at the testing. The core tests all work because of the default settings. New supported formats will alter the default so at least in my head they should also work without alteration.

Alright, thank you.

Two commits:
- Small code style changes
- I used your lang string suggestion. Thank you, your English for sure is better than mine ;-)

**Review by Syxton — COMMENTED** [review]
I understand most of these.  Is there a reason that add_supported_format() has a return void but remove_supported_format() does not?  That coding style is fairly new to me, so I may not fully understand the reasoning.  Thanks

**Comment by PhMemmel** [comment]
@Syxton Yes, reason is I forgot, sorry :-)

**Comment by Syxton** [comment]
@PhMemmel This one is on me.  Just took another look.  I'm not sure what I was thinking in the plugin loops of both functions.  I don't think the isset($name) check is necessary at all.  In my head I was asking "if it is installed" AND.... but the fact that it is in the loop means it has already past that check.

**Comment by PhMemmel** [comment]
> @PhMemmel This one is on me. Just took another look. I'm not sure what I was thinking in the plugin loops of both functions. I don't think the isset($name) check is necessary at all. In my head I was asking "if it is installed" AND.... but the fact that it is in the loop means it has already past that check.

Well, there is a very small case which should not really happen: `isset($name)` is checking if we were able to get the a lang string. That theoretically could happen if the caches have not been purged after a course format installation or something?

**Comment by Syxton** [comment]
@PhMemmel I'm ready to merge this in.  Just wanted to make sure you were at a stopping point.

**Comment by PhMemmel** [comment]
Yes, I'm done so far. I could not test the update methods, have you tested them? Do we maybe want to write at least a unit test for them to make sure they do what they are supposed to? Or are we fine if they eventually fail, because they won't do too much harm? :-)

**Comment by Syxton** [comment]
@PhMemmel I have tested them, and then stripped out the update, then had to add back in a dummy update function to pass checks. :) ok merging in!

**Comment by PhMemmel** [comment]
@Syxton I had one more thought (sorry that it's a bit late): We are using the function names `add_supported_format` and `remove_supported_format` in `upgrade.php` without any namespace. These names aren't very specific to our plugin, so there is a chance we could have a name conflict at some point.

We should consider to rename these functions to `block_massaction_add_supported_format`/`block_massaction_remove_supported_format` or move these methods into the `massactionutils` class or to a separate class as static functions where we can call them safely by the correct namespaced name.

**Comment by Syxton** [comment]
@PhMemmel I agree. I can make those changes since I have to change the pull requests for all the plugins too.

**Comment by PhMemmel** [comment]
Yes, realized that, sorry for the additional work

**Pull request merged**